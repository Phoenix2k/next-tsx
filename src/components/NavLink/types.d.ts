import { ReactNode } from 'react';

export interface NavLinkProps {
  children?: ReactNode;
  href: string;
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;
}
