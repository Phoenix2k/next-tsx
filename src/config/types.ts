export interface Config {
  a11yEnv: Array<string>;
  pages: Pages;
  theme: Theme;
}

export interface Page {
  navLabel: string;
  slug: string;
  title: string;
  url: string;
}

export interface Pages {
  [key: string]: Page;
}

export interface Theme {
  classes: {
    navLinkActive: string;
  };
  colors: {
    backgroundColor: string;
    borderColor: string;
    color: string;
    navLink: string;
    navLinkActive: string;
  };
  typography: {
    sansSerif: string;
  };
}
