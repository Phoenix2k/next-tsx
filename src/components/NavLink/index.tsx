import classnames from 'classnames';
import { useRouter } from 'next/router';
import { NavLinkProps } from './types';

function NavLink({ children, href }: NavLinkProps): JSX.Element {
  const router = useRouter();
  const isActive = router.pathname === href;
  const linkClasses = classnames('link', {
    'nav-link--active': isActive
  });
  const handleClick = (event: React.MouseEvent<HTMLElement>): void => {
    event.preventDefault();
    router.push(href);
  };
  return (
    <a className={linkClasses} href={href} onClick={handleClick}>
      {children}
    </a>
  );
}

export default NavLink;
