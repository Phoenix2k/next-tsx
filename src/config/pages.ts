import sitemap from '../../data/sitemap.json';
import { Pages } from './types';

const pages: Pages = sitemap;

export default pages;
