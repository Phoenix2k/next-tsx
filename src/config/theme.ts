import { Theme } from './types';

const theme: Theme = {
  classes: {
    navLinkActive: 'nav-link--active'
  },
  colors: {
    backgroundColor: 'rgb(255,255,255)',
    borderColor: '#ccc',
    color: 'black',
    navLink: 'blue',
    navLinkActive: 'rgb(0, 128, 0)'
  },
  typography: {
    sansSerif: '"Open Sans", Helvetica, Arial, sans-serif'
  }
};

export default theme;
