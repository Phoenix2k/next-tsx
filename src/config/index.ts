import pages from './pages';
import theme from './theme';
import { Config } from './types';

const config: Config = {
  a11yEnv: ['development'],
  pages,
  theme
};

export default config;
