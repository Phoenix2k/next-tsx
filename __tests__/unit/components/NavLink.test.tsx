/// <reference types="jest"/>

import { ThemeProvider } from 'emotion-theming';
import { RouterContext } from 'next/dist/next-server/lib/router-context';
import { NextRouter } from 'next/router';
import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import Renderer from 'react-test-renderer';
import NavLink from '../../../src/components/NavLink';
import config from '../../../src/config';

describe('Navigation link', () => {
  let container: HTMLElement | null | undefined;

  const mockCallBack = jest.fn();
  const mockEvent = { preventDefault: jest.fn() };
  const mockRouter: Partial<NextRouter> = {
    push: (url: string) => Promise.resolve(!!url)
  };

  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  it('renders and functions correctly', () => {
    act(() => {
      ReactDOM.render(
        // @ts-ignore: Test push method only
        <RouterContext.Provider value={mockRouter}>
          <NavLink href="/test-link" onClick={mockCallBack(mockEvent)}>
            Test link
          </NavLink>
        </RouterContext.Provider>,
        container
      );
    });
    const link: HTMLAnchorElement = container.querySelector('.nav-link');
    expect(link).toBeDefined();
    expect(link.href).toBe('/test-link');
    expect(link.textContent).toBe('Test link');
    act(() => {
      link.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    expect(mockCallBack).toHaveBeenCalledTimes(1);
  });

  it('matches snapshot', () => {
    const snapshotRender = Renderer.create(
      // @ts-ignore: Test push method only
      <RouterContext.Provider value={mockRouter}>
        <ThemeProvider theme={config.theme}>
          <NavLink href="/test-link" onClick={mockCallBack(mockEvent)}>
            Test link
          </NavLink>
        </ThemeProvider>
      </RouterContext.Provider>
    );
    const tree = snapshotRender.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
