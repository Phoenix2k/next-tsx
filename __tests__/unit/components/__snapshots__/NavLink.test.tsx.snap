// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation link matches snapshot 1`] = `
<a
  className="nav-link"
  href="/test-link"
  onClick={[Function]}
>
  Test link
</a>
`;
